File: as_ios/Astra Staging Portal/Astra Staging Portal/DictationController.swift
//
//  DictationController.swift
//  Astra Staging Portal
//
//  App-wide singleton that owns the voice-dictation lifecycle:
//
//    idle ──tap mic──▶ recording ──tap mic──▶ (enqueue + drain)
//
//  The controller holds a single AudioRecorder so recording survives tab
//  switches (user can tap mic on Consultation and keep moving). When the
//  user taps stop, the file is moved into DictationStore and an upload
//  loop starts — if offline, the record sits in the queue until the
//  network comes back. On successful upload, the resulting
//  `DictationRecord` is published via `pendingReview` so the Consultation
//  view can present the review sheet.
//

import Foundation
import Observation

@Observable
@MainActor
final class DictationController {
    static let shared = DictationController()

    enum Phase: Equatable {
        case idle
        case recording(startedAt: Date, stagingId: String, areaId: String?)
        case uploading(clientId: String, stagingId: String)
    }

    private(set) var phase: Phase = .idle
    private(set) var elapsed: TimeInterval = 0
    /// Set when an upload completes — the Consultation view observes this
    /// and presents the review sheet, then clears it.
    var pendingReview: DictationRecord?
    var lastError: String?

    // Context captured when recording starts (used when building the store
    // entry after stop — staging object is captured, not re-read).
    private var currentStaging: Staging?
    private var currentArea: StagingArea?

    // Token + network are supplied by the owning view each frame because
    // the controller doesn't have direct access to either without coupling.
    var authTokenProvider: () -> String? = { nil }

    private let recorder = AudioRecorder()
    private let store = DictationStore.shared
    private let network = NetworkMonitor.shared

    private var tickTimer: Timer?
    private var draining = false

    // MARK: - Recording

    func isRecording(for stagingId: String? = nil) -> Bool {
        switch phase {
        case .recording(_, let sid, _):
            return stagingId == nil || sid == stagingId
        default:
            return false
        }
    }

    /// Ensure we have mic permission, then begin recording. Returns false
    /// if the mic is denied (caller should surface a prompt to open Settings).
    func requestPermissions() async -> (mic: Bool, speech: Bool) {
        await recorder.requestPermissions()
    }

    /// Start a recording for the given staging/area. Call only after
    /// `requestPermissions()` has confirmed mic access.
    func start(staging: Staging, area: StagingArea?) throws {
        guard case .idle = phase else { return }
        try recorder.start()
        currentStaging = staging
        currentArea = area
        phase = .recording(startedAt: Date(), stagingId: staging.id, areaId: area?.id)
        elapsed = 0

        tickTimer?.invalidate()
        tickTimer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                guard let self else { return }
                if case .recording(let startedAt, _, _) = self.phase {
                    self.elapsed = Date().timeIntervalSince(startedAt)
                }
            }
        }
    }

    /// Stop recording, move the audio into the DictationStore, and try to
    /// drain the queue immediately (upload if online).
    func stopAndEnqueue() {
        guard case .recording = phase else { return }
        tickTimer?.invalidate(); tickTimer = nil

        recorder.stop()
        let duration = elapsed
        guard
            case .finished(let fileURL) = recorder.state,
            let staging = currentStaging
        else {
            phase = .idle
            lastError = "Recording ended without a file."
            return
        }

        do {
            let pending = try store.enqueue(
                stagingId: staging.id,
                stagingDisplayName: staging.address ?? staging.name,
                areaId: currentArea?.id,
                areaName: currentArea?.name,
                durationSec: duration,
                sourceFileURL: fileURL,
            )
            phase = .idle
            elapsed = 0
            currentStaging = nil
            currentArea = nil
            drainIfPossible(preferredId: pending.id)
        } catch {
            lastError = "Could not save recording: \(error.localizedDescription)"
            phase = .idle
        }
    }

    /// Cancel a recording without saving. Used when the app backgrounds or
    /// the user explicitly discards.
    func cancel() {
        tickTimer?.invalidate(); tickTimer = nil
        recorder.cancel()
        currentStaging = nil
        currentArea = nil
        elapsed = 0
        phase = .idle
    }

    // MARK: - Upload queue

    /// Kick the drain loop if we're online and not already draining.
    /// `preferredId` lets the caller prioritise a just-enqueued record so
    /// the review sheet pops for the dictation the user just finished.
    func drainIfPossible(preferredId: String? = nil) {
        guard !draining else { return }
        guard network.isOnline else { return }
        guard let token = authTokenProvider() else { return }
        guard store.nextPending() != nil else { return }
        Task { await self.drain(token: token, preferredId: preferredId) }
    }

    private func drain(token: String, preferredId: String?) async {
        draining = true
        defer { draining = false }

        // On the first pass, try to upload `preferredId` first if present,
        // so the review sheet lines up with the user's latest recording.
        if let preferredId,
           let first = store.pending.first(where: { $0.id == preferredId })
        {
            await uploadOne(first, token: token, isPreferred: true)
        }

        while network.isOnline, let next = store.nextPending() {
            await uploadOne(next, token: token, isPreferred: false)
        }
    }

    private func uploadOne(_ record: PendingDictation, token: String, isPreferred: Bool) async {
        var working = record
        working.status = .uploading
        working.attemptCount += 1
        working.lastError = nil
        store.update(working)
        phase = .uploading(clientId: working.id, stagingId: working.stagingId)

        do {
            let resp = try await APIClient.shared.uploadDictation(
                stagingId: working.stagingId,
                areaId: working.areaId,
                areaName: working.areaName,
                clientId: working.id,
                durationSec: working.durationSec,
                fileURL: store.fileURL(for: working),
                token: token,
            )
            store.remove(working)
            phase = .idle
            // Only auto-present the review sheet if this was the preferred
            // upload (the one the user just finished). Queued uploads from
            // an offline session shouldn't hijack the UI.
            if isPreferred {
                pendingReview = resp.dictation
            }
        } catch {
            working.status = .failed
            working.lastError = error.localizedDescription
            store.update(working)
            phase = .idle
            lastError = "Upload failed — will retry when online."
            // Back off briefly so we don't hammer a flapping server.
            try? await Task.sleep(nanoseconds: 4_000_000_000)
        }
    }
}

File: as_android/AstraStagingPortal/app/src/main/java/com/astrastaging/portal/ui/consultation/DictationController.kt
package com.astrastaging.portal.ui.consultation

import android.content.Context
import android.util.Log
import com.astrastaging.portal.data.ApiClient
import com.astrastaging.portal.data.DictationRecord
import com.astrastaging.portal.data.Staging
import com.astrastaging.portal.data.StagingArea
import com.astrastaging.portal.data.network.NetworkMonitor
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * App-wide singleton owning the voice-dictation lifecycle:
 *
 *   idle ──tap mic──▶ recording ──tap mic──▶ (enqueue + drain)
 *
 * The controller holds a single AudioRecorder so recording survives tab
 * switches. When the user stops, the file is moved into DictationStore
 * and an upload loop drains the queue. If offline, the record stays
 * queued until the network comes back. On successful upload, the
 * resulting DictationRecord is published via `pendingReview` so the
 * Consultation view can auto-present the review sheet.
 */
class DictationController(context: Context) {

    sealed class Phase {
        data object Idle : Phase()
        data class Recording(val stagingId: String, val areaId: String?) : Phase()
        data class Uploading(val clientId: String, val stagingId: String) : Phase()
    }

    data class State(
        val phase: Phase = Phase.Idle,
        val elapsedMs: Long = 0,
        val lastError: String? = null,
        /** When an upload completes from a freshly-stopped recording, the
         *  review sheet is presented from this. The view nulls it out. */
        val pendingReview: DictationRecord? = null,
    )

    private val appContext = context.applicationContext
    private val store = DictationStore.get(appContext)
    private val recorder = AudioRecorder(appContext)
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private val drainLock = Mutex()

    private val _state = MutableStateFlow(State())
    val state: StateFlow<State> = _state.asStateFlow()

    // Supplied by the owning view so the controller can fetch the token
    // lazily on upload without having to observe AuthStore directly.
    @Volatile var tokenProvider: () -> String? = { null }

    private var currentStaging: Staging? = null
    private var currentArea: StagingArea? = null
    private var tickJob: Job? = null
    private var recordingStartedAt: Long = 0

    // MARK: - Recording

    fun isRecording(stagingId: String? = null): Boolean {
        val p = _state.value.phase
        return p is Phase.Recording && (stagingId == null || p.stagingId == stagingId)
    }

    fun start(staging: Staging, area: StagingArea?) {
        if (_state.value.phase != Phase.Idle) return
        recorder.start(scope)
        currentStaging = staging
        currentArea = area
        recordingStartedAt = System.currentTimeMillis()
        _state.value = _state.value.copy(
            phase = Phase.Recording(staging.id, area?.id),
            elapsedMs = 0,
            lastError = null,
        )
        tickJob?.cancel()
        tickJob = scope.launch(Dispatchers.Main) {
            while (_state.value.phase is Phase.Recording) {
                val elapsed = System.currentTimeMillis() - recordingStartedAt
                _state.value = _state.value.copy(elapsedMs = elapsed)
                delay(250)
            }
        }
    }

    /** Stop the recording, move the audio into DictationStore, and try
     *  to drain the queue immediately (upload if online). */
    fun stopAndEnqueue() {
        if (_state.value.phase !is Phase.Recording) return
        tickJob?.cancel(); tickJob = null
        recorder.stop()
        val file = recorder.state.value.outputFile
        val staging = currentStaging
        val duration = _state.value.elapsedMs / 1000.0

        if (file == null || staging == null) {
            _state.value = _state.value.copy(
                phase = Phase.Idle,
                lastError = "Recording ended without a file.",
                elapsedMs = 0,
            )
            currentStaging = null
            currentArea = null
            return
        }

        val pending = try {
            store.enqueue(
                stagingId = staging.id,
                stagingDisplayName = staging.address ?: staging.name,
                areaId = currentArea?.id,
                areaName = currentArea?.name,
                durationSec = duration,
                sourceFile = file,
            )
        } catch (t: Throwable) {
            Log.w("DictationController", "enqueue failed", t)
            _state.value = _state.value.copy(
                phase = Phase.Idle,
                lastError = "Could not save recording: ${t.message}",
                elapsedMs = 0,
            )
            currentStaging = null; currentArea = null
            return
        }
        recorder.reset()
        currentStaging = null
        currentArea = null
        _state.value = _state.value.copy(phase = Phase.Idle, elapsedMs = 0)
        drainIfPossible(preferredId = pending.id)
    }

    fun cancel() {
        tickJob?.cancel(); tickJob = null
        recorder.cancel()
        currentStaging = null
        currentArea = null
        _state.value = _state.value.copy(phase = Phase.Idle, elapsedMs = 0)
    }

    /** Called by the view after it consumes a pendingReview value. */
    fun clearPendingReview() {
        if (_state.value.pendingReview != null) {
            _state.value = _state.value.copy(pendingReview = null)
        }
    }

    // MARK: - Upload queue

    fun drainIfPossible(preferredId: String? = null) {
        val token = tokenProvider() ?: return
        if (!NetworkMonitor.snapshot.value.isOnline) return
        if (store.nextPending() == null) return

        scope.launch {
            if (!drainLock.tryLock()) return@launch
            try {
                preferredId?.let { id ->
                    store.load().firstOrNull { it.id == id }?.let { uploadOne(it, token, preferred = true) }
                }
                while (NetworkMonitor.snapshot.value.isOnline) {
                    val next = store.nextPending() ?: break
                    uploadOne(next, token, preferred = false)
                }
            } finally {
                drainLock.unlock()
            }
        }
    }

    private suspend fun uploadOne(record: PendingDictation, token: String, preferred: Boolean) {
        val working = record.copy(
            status = PendingDictationStatus.UPLOADING,
            attemptCount = record.attemptCount + 1,
            lastError = null,
        )
        store.update(working)
        _state.value = _state.value.copy(
            phase = Phase.Uploading(working.id, working.stagingId),
        )

        try {
            val resp = ApiClient.uploadDictation(
                stagingId = working.stagingId,
                areaId = working.areaId,
                areaName = working.areaName,
                clientId = working.id,
                durationSec = working.durationSec,
                file = store.fileFor(working),
                token = token,
            )
            store.remove(working)
            _state.value = _state.value.copy(
                phase = Phase.Idle,
                // Only auto-present for the freshly-stopped recording. A
                // queue drain from an offline session shouldn't hijack UI.
                pendingReview = if (preferred) resp.dictation else _state.value.pendingReview,
            )
        } catch (t: Throwable) {
            Log.w("DictationController", "upload failed", t)
            store.update(working.copy(
                status = PendingDictationStatus.FAILED,
                lastError = t.message ?: "Upload failed",
            ))
            _state.value = _state.value.copy(
                phase = Phase.Idle,
                lastError = "Upload failed — will retry when online.",
            )
            delay(4_000)
        }
    }

    companion object {
        @Volatile
        private var instance: DictationController? = null

        fun get(context: Context): DictationController =
            instance ?: synchronized(this) {
                instance ?: DictationController(context.applicationContext).also { instance = it }
            }
    }
}
